File: backend/app/api/recipe_groups.py
"""app/api/recipe_groups.py

FastAPI router for recipe group endpoints.
Handles recipe group CRUD operations and recipe assignments.
"""

from typing import List

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from app.api.dependencies import get_current_user
from app.database.db import get_session
from app.dtos.recipe_group_dtos import (
    RecipeGroupAssignmentDTO,
    RecipeGroupCreateDTO,
    RecipeGroupResponseDTO,
    RecipeGroupUpdateDTO,
)
from app.models.user import User
from app.services.recipe_group_service import (
    DuplicateRecipeGroupError,
    RecipeGroupNotFoundError,
    RecipeGroupSaveError,
    RecipeGroupService,
)

router = APIRouter()


@router.get("", response_model=List[RecipeGroupResponseDTO])
def list_groups(
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_user),
):
    """
    List all recipe groups for the current user.

    Returns groups ordered alphabetically by name with recipe counts.
    """
    service = RecipeGroupService(session, current_user.id)
    return service.get_all_groups()


@router.get("/{group_id}", response_model=RecipeGroupResponseDTO)
def get_group(
    group_id: int,
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_user),
):
    """Get a single recipe group by ID."""
    service = RecipeGroupService(session, current_user.id)
    try:
        return service.get_group_by_id(group_id)
    except RecipeGroupNotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e))


@router.post("", response_model=RecipeGroupResponseDTO, status_code=201)
def create_group(
    group_data: RecipeGroupCreateDTO,
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_user),
):
    """
    Create a new recipe group.

    - name: Required, 1-255 characters, must be unique per user (case-insensitive)
    """
    service = RecipeGroupService(session, current_user.id)
    try:
        return service.create_group(group_data)
    except DuplicateRecipeGroupError as e:
        raise HTTPException(status_code=409, detail=str(e))
    except RecipeGroupSaveError as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.put("/{group_id}", response_model=RecipeGroupResponseDTO)
def update_group(
    group_id: int,
    update_data: RecipeGroupUpdateDTO,
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_user),
):
    """
    Update a recipe group's name.

    - name: Required, 1-255 characters, must be unique per user (case-insensitive)
    """
    service = RecipeGroupService(session, current_user.id)
    try:
        return service.update_group(group_id, update_data)
    except RecipeGroupNotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except DuplicateRecipeGroupError as e:
        raise HTTPException(status_code=409, detail=str(e))
    except RecipeGroupSaveError as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.delete("/{group_id}")
def delete_group(
    group_id: int,
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_user),
):
    """
    Delete a recipe group.

    Note: This only deletes the group, not the recipes within it.
    """
    service = RecipeGroupService(session, current_user.id)
    try:
        service.delete_group(group_id)
        return {"message": "Recipe group deleted successfully"}
    except RecipeGroupNotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except RecipeGroupSaveError as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/by-recipe/{recipe_id}", response_model=List[RecipeGroupResponseDTO])
def get_groups_for_recipe(
    recipe_id: int,
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_user),
):
    """Get all groups that contain a specific recipe."""
    service = RecipeGroupService(session, current_user.id)
    return service.get_groups_for_recipe(recipe_id)


@router.put("/by-recipe/{recipe_id}/assign", response_model=List[RecipeGroupResponseDTO])
def assign_recipe_to_groups(
    recipe_id: int,
    assignment_data: RecipeGroupAssignmentDTO,
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_user),
):
    """
    Assign a recipe to the specified groups (replaces existing assignments).

    - group_ids: List of group IDs to assign the recipe to
    """
    service = RecipeGroupService(session, current_user.id)
    try:
        return service.assign_recipe_to_groups(recipe_id, assignment_data)
    except RecipeGroupSaveError as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/{group_id}/recipes/{recipe_id}", response_model=RecipeGroupResponseDTO)
def add_recipe_to_group(
    group_id: int,
    recipe_id: int,
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_user),
):
    """Add a recipe to a group."""
    service = RecipeGroupService(session, current_user.id)
    try:
        return service.add_recipe_to_group(group_id, recipe_id)
    except RecipeGroupNotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except RecipeGroupSaveError as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.delete("/{group_id}/recipes/{recipe_id}", response_model=RecipeGroupResponseDTO)
def remove_recipe_from_group(
    group_id: int,
    recipe_id: int,
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_user),
):
    """Remove a recipe from a group."""
    service = RecipeGroupService(session, current_user.id)
    try:
        return service.remove_recipe_from_group(group_id, recipe_id)
    except RecipeGroupNotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except RecipeGroupSaveError as e:
        raise HTTPException(status_code=500, detail=str(e))

File: frontend/src/lib/api/recipe-groups.ts
import type {
  RecipeGroupResponseDTO,
  RecipeGroupCreateDTO,
  RecipeGroupUpdateDTO,
  RecipeGroupAssignmentDTO,
} from "@/types/recipe";
import { fetchApi } from "./base";

export const recipeGroupApi = {
  /**
   * List all recipe groups for the current user
   * @param token - Auth token for authenticated requests
   * @returns Array of recipe groups with recipe counts
   */
  list: (token: string | null): Promise<RecipeGroupResponseDTO[]> =>
    fetchApi<RecipeGroupResponseDTO[]>("/api/recipe-groups", undefined, token),

  /**
   * Get a single recipe group by ID
   * @param id - Group ID
   * @param token - Auth token for authenticated requests
   * @returns Recipe group details
   */
  get: (id: number, token: string | null): Promise<RecipeGroupResponseDTO> =>
    fetchApi<RecipeGroupResponseDTO>(`/api/recipe-groups/${id}`, undefined, token),

  /**
   * Create a new recipe group
   * @param data - Group data (name)
   * @param token - Auth token for authenticated requests
   * @returns Created recipe group
   */
  create: (data: RecipeGroupCreateDTO, token: string | null): Promise<RecipeGroupResponseDTO> =>
    fetchApi<RecipeGroupResponseDTO>(
      "/api/recipe-groups",
      {
        method: "POST",
        body: JSON.stringify(data),
      },
      token
    ),

  /**
   * Update a recipe group's name
   * @param id - Group ID
   * @param data - Updated group data
   * @param token - Auth token for authenticated requests
   * @returns Updated recipe group
   */
  update: (id: number, data: RecipeGroupUpdateDTO, token: string | null): Promise<RecipeGroupResponseDTO> =>
    fetchApi<RecipeGroupResponseDTO>(
      `/api/recipe-groups/${id}`,
      {
        method: "PUT",
        body: JSON.stringify(data),
      },
      token
    ),

  /**
   * Delete a recipe group
   * @param id - Group ID
   * @param token - Auth token for authenticated requests
   */
  delete: (id: number, token: string | null): Promise<void> =>
    fetchApi<void>(
      `/api/recipe-groups/${id}`,
      {
        method: "DELETE",
      },
      token
    ),

  /**
   * Get all groups that contain a specific recipe
   * @param recipeId - Recipe ID
   * @param token - Auth token for authenticated requests
   * @returns Array of groups containing this recipe
   */
  getGroupsForRecipe: (recipeId: number, token: string | null): Promise<RecipeGroupResponseDTO[]> =>
    fetchApi<RecipeGroupResponseDTO[]>(
      `/api/recipe-groups/by-recipe/${recipeId}`,
      undefined,
      token
    ),

  /**
   * Assign a recipe to specific groups (replaces existing assignments)
   * @param recipeId - Recipe ID
   * @param data - Group IDs to assign
   * @param token - Auth token for authenticated requests
   * @returns Updated list of groups for the recipe
   */
  assignRecipeToGroups: (
    recipeId: number,
    data: RecipeGroupAssignmentDTO,
    token: string | null
  ): Promise<RecipeGroupResponseDTO[]> =>
    fetchApi<RecipeGroupResponseDTO[]>(
      `/api/recipe-groups/by-recipe/${recipeId}/assign`,
      {
        method: "PUT",
        body: JSON.stringify(data),
      },
      token
    ),

  /**
   * Add a recipe to a group
   * @param groupId - Group ID
   * @param recipeId - Recipe ID
   * @param token - Auth token for authenticated requests
   * @returns Updated recipe group
   */
  addRecipeToGroup: (
    groupId: number,
    recipeId: number,
    token: string | null
  ): Promise<RecipeGroupResponseDTO> =>
    fetchApi<RecipeGroupResponseDTO>(
      `/api/recipe-groups/${groupId}/recipes/${recipeId}`,
      {
        method: "POST",
      },
      token
    ),

  /**
   * Remove a recipe from a group
   * @param groupId - Group ID
   * @param recipeId - Recipe ID
   * @param token - Auth token for authenticated requests
   * @returns Updated recipe group
   */
  removeRecipeFromGroup: (
    groupId: number,
    recipeId: number,
    token: string | null
  ): Promise<RecipeGroupResponseDTO> =>
    fetchApi<RecipeGroupResponseDTO>(
      `/api/recipe-groups/${groupId}/recipes/${recipeId}`,
      {
        method: "DELETE",
      },
      token
    ),
};
